### PR TITLE
gh #24 deepsleep: L3 Test case Development

### PR DESCRIPTION
**Pull request by hgfell683**
Create the python unit test using the ut_raft and verify that they are working properly.

### CODE REVIEW

**Review by bhanucbp — COMMENTED** [review]
address review comments

File: host/tests/deepsleep_L3_Tests/triggerDeepsleep.py [hunk 1]
```diff
@@ -0,0 +1,315 @@
...
+
+    testName  = "test1_TriggerDeepsleep"
+    testSetupPathPower = dir_path + "/power_L3_testSetup.yml"
+    testSetupPathDeepsleep = dir_path + "/deepsleep_L3_testSetup.yml"
+     = dir_path + "../../../profiles/deepsleepmanagerExtendedEnumsNotSupported.yml"
+    moduleNameDeepsleep = "deepsleepmanager"
+    moduleNamePower = "powermanager"
+    rackDevice = "dut"
+
+    wakeup_sources_map = {
```
[the remark this answers is not in the document]
Comment (hgfell683) [reply host/tests/deepsleep_L3_Tests/triggerDeepsleep.py:50]: Don't we still want to have some 'map' to reference the names for logging purposes of the enum values?

File: host/tests/deepsleepClasses/deepsleep.py [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+dir_path = os.path.dirname(os.path.realpath(__file__))
+sys.path.append(dir_path+"/../")
+
+from raft.framework.plugins.ut_raft.configRead import ConfigRead
+from raft.framework.plugins.ut_raft.utSuiteNavigator import UTSuiteNavigatorClass
+from raft.framework.plugins.ut_raft.interactiveShell import InteractiveShell
+
+class deepsleepClass():
+
+    moduleName = "deepsleep"
```
[the remark this answers is not in the document]
Comment (hgfell683) [reply host/tests/deepsleepClasses/deepsleep.py]: SHould be updated now.

File: host/tests/deepsleepClasses/deepsleep.py [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+            },
+            {
+                "query_type": "direct",
+                "query": "Network Standby enabled/disabled [0/1]",
+                "input": "1"
+            }
+        ]
+
+        # Convert input arguments to strings and update the prompts
+        promptWithAnswers[0]["input"] = str(timeout)
```
[the remark this answers is not in the document]
Comment (hgfell683) [reply host/tests/deepsleepClasses/deepsleep.py]: SHould be updated now.

File: host/tests/deepsleepClasses/deepsleep.py [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+        """
+        Get last wakeup reason
+
+        Args:
+            None
+
+        Returns:
+            None
+        """
+        result = self.utMenu.select( self.testSuite, "Get last wakeup reason")
```
[the remark this answers is not in the document]
Comment (hgfell683) [reply host/tests/deepsleepClasses/deepsleep.py:142]: SHould be updated now.

File: host/tests/deepsleepClasses/deepsleep.py [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+        """
+        Get last wakeup keycode.
+
+        Args:
+            None
+
+        Returns:
+            None
+        """
+        result = self.utMenu.select( self.testSuite, "Get last wakeup keycode")
```
[the remark this answers is not in the document]
Comment (hgfell683) [reply host/tests/deepsleepClasses/deepsleep.py:165]: SHould be updated now.

File: host/tests/deepsleep_L3_Tests/triggerDeepsleep.py [hunk 6]
```diff
@@ -0,0 +1,315 @@
+#!/usr/bin/env python3
```
[the remark this answers is not in the document]
Comment (hgfell683) [reply host/tests/deepsleep_L3_Tests/triggerDeepsleep.py]: Should be updated now.

File: host/tests/deepsleep_L3_Tests/triggerDeepsleep.py [hunk 7]
```diff
@@ -0,0 +1,315 @@
...
+dir_path = os.path.dirname(os.path.realpath(__file__))
+sys.path.append(dir_path+"/../")
+
+from deepsleepClases.deepsleep import deepsleepClass
+from deepsleepCltestSetupProfilesases.power import powerManagerClass
+from raft.framework.plugins.ut_raft import utHelperClass
+from raft.framework.plugins.ut_raft.configRead import ConfigRead
+from raft.framework.plugins.ut_raft.utUserResponse import utUserResponse
+
+class dsHost_test1_TriggerDeepsleep(utHelperClass):
```
[the remark this answers is not in the document]
Comment (hgfell683) [reply host/tests/deepsleep_L3_Tests/triggerDeepsleep.py]: Should be updated now.

File: host/tests/deepsleep_L3_Tests/triggerDeepsleep.py [hunk 8]
```diff
@@ -0,0 +1,315 @@
...
+
+    testName  = "test1_TriggerDeepsleep"
+    testSetupPathPower = dir_path + "/power_L3_testSetup.yml"
+    testSetupPathDeepsleep = dir_path + "/deepsleep_L3_testSetup.yml"
+     = dir_path + "../../../profiles/deepsleepmanagerExtendedEnumsNotSupported.yml"
```
[the remark this answers is not in the document]
Comment (hgfell683) [reply host/tests/deepsleep_L3_Tests/triggerDeepsleep.py]: Should be updated now.

File: host/tests/deepsleep_L3_Tests/triggerDeepsleep.py [hunk 9]
```diff
@@ -0,0 +1,315 @@
...
+        Args:
+            None.
+        """
+        super().__init__(self.testName, '1')
+
+        # Test Setup configuration file
+        self.testSetupPower = ConfigRead(self.testSetupPathPower, self.moduleNamePower)
+        self.testSetupDeepsleep = ConfigRead(self.testSetupPathDeepsleep, self.moduleNameDeepsleep)
+
+        self.wakeupsources = ConfigRead(self.testSetupProfiles, "powermanager")
```
[the remark this answers is not in the document]
Comment (hgfell683) [reply host/tests/deepsleep_L3_Tests/triggerDeepsleep.py]: Should be updated now.

File: host/tests/deepsleep_L3_Tests/triggerDeepsleep.py [hunk 10]
```diff
@@ -0,0 +1,315 @@
...
+
+
+        # Create the Power Manager Class
+        self.testPower = powerManagerClass(self.deviceProfile, self.hal_session_power)
+
+        # Initialize the deepsleep manager and power manager
+        self.testPower.initialise()
+        self.testDeepsleep.initialise()
+
+        wakeupsources = self.wakeupsources.WakeupSources #hex value
```
[the remark this answers is not in the document]
Comment (hgfell683) [reply host/tests/deepsleep_L3_Tests/triggerDeepsleep.py]: Should be updated now.

File: host/tests/deepsleep_L3_Tests/triggerDeepsleep.py [hunk 11]
```diff
@@ -0,0 +1,315 @@
...
+
+        for source in self.wakeup_sources_map:
```
[the remark this answers is not in the document]
Comment (hgfell683) [reply host/tests/deepsleep_L3_Tests/triggerDeepsleep.py]: Should be updated now.

File: host/tests/deepsleep_L3_Tests/triggerDeepsleep.py [hunk 12]
```diff
@@ -0,0 +1,315 @@
...
+
+    def testFunction(self):
+        """This function will test the get cpu tempature functionality
+
+        Returns:
+            bool
+        """
+
+        # Download the assets listed in test setup configuration file
+        self.testDownloadAssets()
```
[the remark this answers is not in the document]
Comment (hgfell683) [reply host/tests/deepsleep_L3_Tests/triggerDeepsleep.py]: Removed

File: host/tests/deepsleep_L3_Tests/triggerDeepsleep.py [hunk 13]
```diff
@@ -0,0 +1,315 @@
...
+        self.testStreams = []
+
+        self.deviceDownloadPath = self.cpe.get("target_directory")
+
+        #download test artifacts to device
+        url = self.testSetup.assets.device.test1_TriggerDeepsleep.artifacts
+        if url is not None:
+            self.downloadToDevice(url, self.deviceDownloadPath, self.rackDevice)
+
+    def testCleanAssets(self):
```
[the remark this answers is not in the document]
Comment (hgfell683) [reply host/tests/deepsleep_L3_Tests/triggerDeepsleep.py]: Removed

File: docs/pages/deep-sleep-manager_L3_TestProcedure.md [hunk 14]
```diff
@@ -0,0 +1,209 @@
...
+  - Select and execute the Python file: **`test1_TriggerDeepsleep.py`**
+  - The test will automatically download all required artifacts, copying them to the designated target directory before commencing execution.
+
+- Trigger deepsleep prompt:
+
+    Would you like to trigger WAKE-UP SOURCE test? Note, skipping the test is considered a fail condition. (Y/N):
+
+    Please trigger wake up via [will be mention whatever wake up source is being tested]. Did the device wakeup as expected (Y/N):
+
+- The above will loop through all wake up sources that acclicable for the specific device based on the .yaml file.
```
Comment (bhanucbp) [inline docs/pages/deep-sleep-manager_L3_TestProcedure.md:209]: Add details of manual triggering of wakeup. How to wakeup device and how to verify

File: host/tests/deepsleepClasses/deepsleep.py [hunk 15]
```diff
@@ -0,0 +1,217 @@
...
+            },
+            {
+                "query_type": "direct",
+                "query": "Network Standby enabled/disabled [0/1]",
+                "input": "1"
+            }
+        ]
+
+        # Convert input arguments to strings and update the prompts
+        promptWithAnswers = [
```
Comment (bhanucbp) [inline host/tests/deepsleepClasses/deepsleep.py:106]: remove duplicate entry

File: host/tests/deepsleepClasses/power.py [hunk 16]
```diff
@@ -0,0 +1,233 @@
...
+
+        Returns:
+            None
+        """
+
+        promptWithAnswers = [
+            {
+                "query_type": "direct",
+                "query": "Power State",
+                "input": "PWRMGR_POWERSTATE_ON"
```
Comment (bhanucbp) [inline host/tests/deepsleepClasses/power.py:117]: directly assign the input here

```
promptWithAnswers = [
            {
                "query_type": "direct",
                "query": "Power State",
                "input": str(mode)
}
]
```

File: host/tests/deepsleepClasses/power.py [hunk 17]
```diff
@@ -0,0 +1,233 @@
...
+        """
+        self.utMenu.stop()
+
+# Test and example usage code
+if __name__ == '__main__':
+
+    shell = InteractiveShell()
+    shell.open()
+
+    platformProfile = "https://github.com/rdkcentral/rdk-halif-test-power_manager/blob/main/profiles/source/source_powerManager.yaml" #Does this work? Do we need something else?
```
Comment (bhanucbp) [inline host/tests/deepsleepClasses/power.py:222]: http wont work
yaml file is part of the test, you can give relative path to the profile file

**Review by bhanucbp — CHANGES_REQUESTED** [review]
address review comments

File: host/tests/deepsleep_L3_Tests/test1_TriggerDeepsleep.py [hunk 18]
```diff
@@ -0,0 +1,337 @@
...
+
+        wakeupsources = self.wakeupsources['powermanager']['WakeupSources'] #hex value
+
+        skipped = False
+
+        result = True
+
+        for source in wakeupsources:
+            if source == 0x06:
+                # Timer
```
[the remark this answers is not in the document]
Comment (hgfell683) [reply host/tests/deepsleep_L3_Tests/test1_TriggerDeepsleep.py:246]: We already test timer in L1 and L2 tests, and there is variation in those tests so wouldn't adding timer be redundant?

File: host/tests/deepsleepClasses/power.py [hunk 19]
```diff
@@ -0,0 +1,276 @@
...
+        Returns:
+            Wakeup sources
+        
+        """
+
+        return ConfigRead(self.getProfile(), "powermanager")['powermanager']['WakeupSources']
+
+
+    def testDownloadAssets(self):
+        """
```
Comment (bhanucbp) [inline host/tests/deepsleepClasses/power.py:216]: remove this function

File: host/tests/deepsleepClasses/power.py [hunk 20]
```diff
@@ -0,0 +1,276 @@
...
+if __name__ == '__main__':
+
+    shell = InteractiveShell()
+    shell.open()
+
+    platformProfile = dir_path + "/../../../profiles/sinkWakeUpSources.yaml" #Does this work? Do we need something else?
+    # test the class assuming that it's optional
+    test = powerManagerClass(platformProfile, shell)
+
+    test.testDownloadAssets()
```
Comment (bhanucbp) [inline host/tests/deepsleepClasses/power.py:267]: remove this line

File: host/tests/deepsleepClasses/power.py [hunk 21]
```diff
@@ -0,0 +1,276 @@
...
+        """
+        Get Power Mode.
+
+        Args:
+            None
+
+        Returns:
+            None
+        """
+        result = self.utMenu.select( self.testSuite, "Get Power State")
```
Comment (bhanucbp) [inline host/tests/deepsleepClasses/power.py:143]: this is a get function, but its not returning the mode

File: host/tests/deepsleepClasses/power_testConfig.yml [hunk 22]
```diff
@@ -0,0 +1,23 @@
+# this file is generated by the testing suites, when a new test is modified, this configuration file will be filled out and updated.
+
+power:  # Prefix must always exist
+    description: "power Manager testing profile / menu system for UT"
+    test:
+        execute: "" #Execute command
```
Comment (bhanucbp) [inline host/tests/deepsleepClasses/power_testConfig.yml:6]: Add artifacts folder and update execute field (refer dsAudio)

File: host/tests/deepsleepClasses/deepsleep_testConfig.yml [hunk 23]
```diff
@@ -15,9 +15,9 @@ deepsleep:  # Prefix must always exist
             2:
```
Comment (bhanucbp) [inline host/tests/deepsleepClasses/deepsleep_testConfig.yml:15]: Add L1 and L2 test cases also

File: host/tests/deepsleepClasses/power_testConfig.yml [hunk 24]
```diff
@@ -0,0 +1,23 @@
...
+        type: UT-C # C (UT-C Cunit) / C++ (UT-G (g++ ut-core gtest backend))
+        suites:
```
Comment (bhanucbp) [inline host/tests/deepsleepClasses/power_testConfig.yml:8]: Add L1 and L2 test cases also

File: host/tests/deepsleep_L3_Tests/test1_TriggerDeepsleep.py [hunk 25]
```diff
@@ -0,0 +1,382 @@
...
+            self.log.step(f"Triggering deepsleep")
+            if source == 0x06:
+                self.testDeepsleep.triggerDeepsleep(60, 1)
+            else:
+                self.testDeepsleep.triggerDeepsleep(0, 1)
+
+            if source == 0x00:
+               # Wake on Voice
+               result = self.wake_on_voice("rtmp://your-backend-server/live/stream", "audio.wav", True)
+               if result:
```
Comment (bhanucbp) [inline host/tests/deepsleep_L3_Tests/test1_TriggerDeepsleep.py:298]: this code is repeating for all the verification, add once after all the elif 

```
 if result:
                   self.log.stepResult(result, f"Voice wakeup success")
               else:
                   self.log.stepResult(result, f"Voice wakeup failed")
               pass

```

replace with

```
self.log.stepStart(f"Triggering wake up test for {self.testPower.wakeup_sources_map[source]}")
if
elif
elif
elif
:
:
 self.log.stepResult(result, f"Wake up test for {self.testPower.wakeup_sources_map[source]}")
```

**Review by bhanucbp — COMMENTED** [review]
address review comments

File: docs/pages/deep-sleep-manager_L3_TestProcedure.md (hunk 14 above)
Comment (hgfell683) [reply docs/pages/deep-sleep-manager_L3_TestProcedure.md]: Updated, with how to trigger, and how to verify.

File: host/tests/deepsleepClasses/deepsleep.py [hunk 26]
```diff
@@ -0,0 +1,228 @@
...
+
+        Returns:
+            None
+        """
+        result = self.utMenu.select( self.testSuite, "Get last wakeup reason")
+
+
+        # Define the pattern to capture the wakeup reason
+        wakeupReasonPattern = r'Result PLAT_DS_GetLastWakeupReason\(wakeupReason:\[(?P<wakeupReason>[^\]]+)\]\)'
+        match = re.search(wakeupReasonPattern, result)
```
[the remark this answers is not in the document]
Comment (hgfell683) [reply host/tests/deepsleepClasses/deepsleep.py]: Removed

File: host/tests/deepsleepClasses/deepsleep.py [hunk 27]
```diff
@@ -0,0 +1,228 @@
...
+            None
+
+        Returns:
+            None
+        """
+        result = self.utMenu.select( self.testSuite, "Get last wakeup keycode")
+
+        # Define the pattern to capture the wakeup keycode
+        wakeupKeycodePattern = r'Result PLAT_DS_GetLastWakeupKeyCode\(wakeupKeyCode:\[(?P<wakeupKeycode>\d+)\]\)'
+        match = re.search(wakeupKeycodePattern, result)
```
[the remark this answers is not in the document]
Comment (hgfell683) [reply host/tests/deepsleepClasses/deepsleep.py]: Removed

File: host/tests/deepsleepClasses/deepsleep.py [hunk 28]
```diff
@@ -0,0 +1,228 @@
...
+if __name__ == '__main__':
+
+    shell = InteractiveShell()
+    shell.open()
+
+    platformProfile = dir_path + "/../../../profiles/deepsleepmanagerExtendedEnumsNotSupported.yaml"
+    # test the class assuming that it's optional
+    test = deepsleepClass(platformProfile, shell)
+
+    test.testDownloadAssets()
```
[the remark this answers is not in the document]
Comment (hgfell683) [reply host/tests/deepsleepClasses/deepsleep.py]: Removed

File: host/tests/deepsleepClasses/power.py [hunk 29]
```diff
@@ -0,0 +1,276 @@
...
+        self.utMenu        = UTSuiteNavigatorClass(self.testConfig, None, session)
+        self.testSession = session
+
+        for artifact in self.testConfig.test.artifacts:
+            filesPath = os.path.join(dir_path, artifact)
+            self.utils.rsync(self.testSession, filesPath, targetWorkspace)
+
+        self.utMenu.start()
+
+    class WakeupSources(Enum):
```
[the remark this answers is not in the document]
Comment (hgfell683) [reply host/tests/deepsleepClasses/power.py]: Moved to outside of the class

File: host/tests/deepsleepClasses/power.py [hunk 30]
```diff
@@ -0,0 +1,276 @@
...
+        promptWithAnswers = [
+            {
+                "query_type": "direct",
+                "query": "Select Wake Up Source Type: ",
+                "input": "PWRMGR_WAKEUPSRC_VOICE"
+            }
+        ]
+
+        # Convert input arguments to strings and update the prompts
+        promptWithAnswers[0]["input"] = str(source)
```
[the remark this answers is not in the document]
Comment (hgfell683) [reply host/tests/deepsleepClasses/power.py]: Updated

File: host/tests/deepsleepClasses/power.py (hunk 19 above)
Comment (hgfell683) [reply host/tests/deepsleepClasses/power.py]: REmoved

File: host/tests/deepsleepClasses/power.py (hunk 20 above)
Comment (hgfell683) [reply host/tests/deepsleepClasses/power.py]: Removed

File: host/tests/deepsleepClasses/power.py (hunk 21 above)
Comment (hgfell683) [reply host/tests/deepsleepClasses/power.py:141]: Updated with a return

File: host/tests/deepsleepClasses/power_testConfig.yml (hunk 22 above)
Comment (hgfell683) [reply host/tests/deepsleepClasses/power_testConfig.yml]: Added as requested

File: host/tests/deepsleepClasses/deepsleep_testConfig.yml (hunk 23 above)
Comment (hgfell683) [reply host/tests/deepsleepClasses/deepsleep_testConfig.yml:33]: Added additional casses

File: host/tests/deepsleepClasses/power_testConfig.yml (hunk 24 above)
Comment (hgfell683) [reply host/tests/deepsleepClasses/power_testConfig.yml:12]: Added addtional cases

File: host/tests/deepsleep_L3_Tests/test1_TriggerDeepsleep.py (hunk 25 above)
Comment (hgfell683) [reply host/tests/deepsleep_L3_Tests/test1_TriggerDeepsleep.py]: Updated

File: host/tests/deepsleep_L3_Tests/triggerDeepsleep.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (hgfell683) [reply host/tests/deepsleep_L3_Tests/triggerDeepsleep.py]: Is this no removed?

**Review by thanushreevani — DISMISSED** [review]

File: docs/pages/deep-sleep-manager_L3_TestProcedure.md [hunk 31]
```diff
@@ -201,20 +216,20 @@ All tested wake-up sources must successfully trigger the system to exit deep sle
...
+- Once in deepsleep the device will need to be manually awoken for some sources like `IR`, `CEC` and `Power_Key`.
 
-    - When the device enters 'IR' test, to trigger the wake up use an IR remote pressing the power key on the remote to wake up the device. If the device successfully wakes up from deepsleep via the specified method, User should enter the response it passes the test.
+    - During the `IR` test, use an `IR` remote to trigger the wake-up by pressing the power key. If the device successfully wakes up from deep sleep using this method, the test evaluates whether the wake-up process was successful.
 
-    - When the device enters 'power key' test, to trigger the wake up press the front panel power key button of the device itself. If the device successfully wakes up from deepsleep via the specified method, User should enter the response it passes the test.
+    - During the `Power_Key` test, press the power key of the `dut` to wake-up. If the device successfully wakes up from deep sleep using this method, the test evaluates whether the wake-up process was successful.
```
Comment (srinivasgtl) [inline docs/pages/deep-sleep-manager_L3_TestProcedure.md:225]: Porbably the IR wakeup keys can be Power and Home Keys.

Comment (bhanucbp) [reply docs/pages/deep-sleep-manager_L3_TestProcedure.md:225]: updated the line with home key

**Review by hari22yuva — APPROVED** [review]
looks ok

**Review by bhanucbp — APPROVED** [review]
ship it

**Pull request merged**